File: core/src/main/java/com/netflix/iceberg/BaseSnapshot.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.netflix.iceberg;

import com.google.common.base.Objects;
import com.google.common.collect.Lists;
import com.netflix.iceberg.exceptions.RuntimeIOException;
import com.netflix.iceberg.io.CloseableGroup;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;

class BaseSnapshot extends CloseableGroup implements Snapshot {
  private final TableOperations ops;
  private final long snapshotId;
  private final Long parentId;
  private final long timestampMillis;
  private final List<String> manifestFiles;

  // lazily initialized
  private List<DataFile> adds = null;
  private List<DataFile> deletes = null;

  /**
   * For testing only.
   */
  BaseSnapshot(TableOperations ops,
               long snapshotId,
               String... manifestFiles) {
    this(ops, snapshotId, null, System.currentTimeMillis(), Arrays.asList(manifestFiles));
  }

  BaseSnapshot(TableOperations ops,
               long snapshotId,
               Long parentId,
               long timestampMillis,
               List<String> manifestFiles) {
    this.ops = ops;
    this.snapshotId = snapshotId;
    this.parentId = parentId;
    this.timestampMillis = timestampMillis;
    this.manifestFiles = manifestFiles;
  }

  @Override
  public long snapshotId() {
    return snapshotId;
  }

  @Override
  public Long parentId() {
    return parentId;
  }

  @Override
  public long timestampMillis() {
    return timestampMillis;
  }

  @Override
  public List<String> manifests() {
    return manifestFiles;
  }

  @Override
  public List<DataFile> addedFiles() {
    if (adds == null) {
      cacheChanges();
    }
    return adds;
  }

  @Override
  public List<DataFile> deletedFiles() {
    if (deletes == null) {
      cacheChanges();
    }
    return deletes;
  }

  private void cacheChanges() {
    List<DataFile> adds = Lists.newArrayList();
    List<DataFile> deletes = Lists.newArrayList();

    // accumulate adds and deletes from all manifests.
    // because manifests can be reused in newer snapshots, filter the changes by snapshot id.
    for (String manifest : manifestFiles) {
      try (ManifestReader reader = ManifestReader.read(ops.newInputFile(manifest))) {
        for (ManifestEntry add : reader.addedFiles()) {
          if (add.snapshotId() == snapshotId) {
            adds.add(add.file().copy());
          }
        }
        for (ManifestEntry delete : reader.deletedFiles()) {
          if (delete.snapshotId() == snapshotId) {
            deletes.add(delete.file().copy());
          }
        }
      } catch (IOException e) {
        throw new RuntimeIOException(e, "Failed to close reader while caching changes");
      }
    }

    this.adds = adds;
    this.deletes = deletes;
  }

  @Override
  public String toString() {
    return Objects.toStringHelper(this)
        .add("id", snapshotId)
        .add("timestamp_ms", timestampMillis)
        .add("manifests", manifestFiles)
        .toString();
  }
}

File: python/iceberg/core/base_snapshot.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import time

from iceberg.api import (Filterable,
                         FilteredSnapshot,
                         ManifestFile,
                         Snapshot,
                         SnapshotIterable)
from iceberg.api.expressions import Expressions
from iceberg.api.io import CloseableGroup
from iceberg.core.avro import AvroToIceberg

from .generic_manifest_file import GenericManifestFile
from .manifest_reader import ManifestReader


class BaseSnapshot(Snapshot, SnapshotIterable, CloseableGroup):

    @staticmethod
    def snapshot_from_files(ops, snapshot_id, files):
        return BaseSnapshot(ops, snapshot_id, None,
                            manifests=[GenericManifestFile(file=ops.new_input_file(path), spec_id=0)
                                       for path in files])

    def __init__(self, ops, snapshot_id, parent_id=None, manifests=None, manifest_list=None, timestamp_millis=None,
                 operation=None, summary=None):
        super(BaseSnapshot, self).__init__()
        if timestamp_millis is None:
            timestamp_millis = int(time.time() * 1000)

        self._ops = ops
        self._snapshot_id = snapshot_id
        self._parent_id = parent_id
        self._timestamp_millis = timestamp_millis
        if manifests is not None:
            self._manifests = [manifest if isinstance(manifest, GenericManifestFile)
                               else GenericManifestFile(file=ops.new_input_file(manifest), spec_id=0)
                               for manifest in manifests]
        else:
            self._manifests = None
        self._manifest_list = manifest_list
        self._operation = operation
        self._summary = summary

        self._adds = None
        self._deletes = None

    @property
    def snapshot_id(self):
        return self._snapshot_id

    @property
    def timestamp_millis(self):
        return self._timestamp_millis

    @property
    def parent_id(self):
        return self._parent_id

    @property
    def manifests(self):
        if self._manifests is None:
            # if manifest isn't set then the snapshot_file is set and should be read to get the list
            return (GenericManifestFile.from_avro_record_json(manifest)
                    for manifest in AvroToIceberg.read_avro_file(ManifestFile.schema(), self._manifest_list))

        return self._manifests

    @property
    def manifest_location(self):
        return self._manifest_list.location if self._manifest_list is not None else None

    @property
    def summary(self):
        return self._summary

    @property
    def operation(self):
        return self._operation

    def select(self, columns):
        return FilteredSnapshot(self, Expressions.always_true(), Expressions.always_true(), columns)

    def filter_partitions(self, expr):
        return FilteredSnapshot(self, expr, Expressions.always_true(), Filterable.ALL_COLUMNS)

    def filter_rows(self, expr):
        return FilteredSnapshot(self, Expressions.always_true(), expr, Filterable.ALL_COLUMNS)

    def iterator(self, part_filter=None, row_filter=None, columns=None):
        if part_filter is None and row_filter is None and columns is None:
            return self.iterator(Expressions.always_true(), Expressions.always_true(), Filterable.ALL_COLUMNS)

        return iter([self.get_filtered_manifest(path, part_filter, row_filter, columns)
                     for path in self._manifest_files])

    def added_files(self):
        raise NotImplementedError()

    def deleted_files(self):
        raise NotImplementedError()

    def cache_changes(self):
        raise NotImplementedError

    def __repr__(self):
        return "BaseSnapshot(id={id},timestamp_ms={ts_ms},manifests={manifests}".format(id=self._snapshot_id,
                                                                                        ts_ms=self._timestamp_millis,
                                                                                        manifests=self._manifests)

    def __str__(self):
        return self.__repr__()

    def get_filtered_manifest(self, path, part_filter, row_filter, columns):
        reader = ManifestReader.read(self.ops.new_input_file(path))
        self.add_closeable(reader)
        return reader
